src/primitives/transaction.h

```diff
@@ -616,21 +655,168 @@ class CTxCertifierLockOut : public CTxCrosschainOut
     }
 };
 
+// forward declarations
+class CValidationState;
+class CTxMemPool;
+class CCoinsViewCache;
+class CChain;
+class CBlock;
+class CBlockTemplate;
+class CScriptCheck;
+class CBlockUndo;
+class CTxUndo;
+class UniValue;
+
+namespace Sidechain { class ScCoinsViewCache; }
+
+class BaseSignatureChecker;
+
+// abstract interface for CTransaction and CScCertificate
+class CTransactionBase
+{
+public:
+    const int32_t nVersion;
+protected:
+    const std::vector<CTxIn> vin;
+    const std::vector<CTxOut> vout;
+
+    /** Memory only. */
+    const uint256 hash;
+
+    virtual void UpdateHash() const = 0;
+
+public:
+    virtual bool TryPushToMempool(bool fLimitFree, bool fRejectAbsurdFee) = 0;
+
+    CTransactionBase();
+    CTransactionBase& operator=(const CTransactionBase& tx);
+    CTransactionBase(const CTransactionBase& tx);
+    virtual ~CTransactionBase() {};
+
+    template <typename Stream>
+    CTransactionBase(deserialize_type, Stream& s) : CTransactionBase(CMutableTransactionBase(deserialize, s)) {}
+
+    const uint256& GetHash() const {
+        return hash;
+    }
+
+    friend bool operator==(const CTransactionBase& a, const CTransactionBase& b)
+    {
+        return a.hash == b.hash;
+    }
+
+    friend bool operator!=(const CTransactionBase& a, const CTransactionBase& b)
+    {
+        return !(a==b);
+    }
+
+    //GETTERS
+    const std::vector<CTxIn>&         GetVin()        const {return vin;};
+    const std::vector<CTxOut>&        GetVout()       const {return vout;};
+
+    virtual const std::vector<CTxScCreationOut>&      GetVscCcOut()   const = 0;
+    virtual const std::vector<CTxCertifierLockOut>&   GetVclCcOut()   const = 0;
+    virtual const std::vector<CTxForwardTransferOut>& GetVftCcOut()   const = 0;
+    virtual const std::vector<JSDescription>&         GetVjoinsplit() const = 0;
+    virtual const uint256&                            GetScId()       const = 0;
+    virtual const uint32_t&                           GetLockTime()   const = 0;
+    //END OF GETTERS
+
+    //CHECK FUNCTIONS
+    virtual bool CheckVersionBasic        (CValidationState &state) const = 0;
+    virtual bool CheckVersionIsStandard   (std::string& reason, int nHeight) const = 0;
+    virtual bool CheckInputsAvailability  (CValidationState &state) const = 0;
+    virtual bool CheckOutputsAvailability (CValidationState &state) const = 0;
+    virtual bool CheckSerializedSize      (CValidationState &state) const = 0;
+
+    bool CheckInputsAmount (CValidationState &state) const;
```

Comment:
aggregate with CheckOutputsAmount and rename to CheckAmounts

Reply:
Done